qiskit/algorithms/minimum_eigen_solvers/adaptvqe2.py

```diff
@@ -0,0 +1,353 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A ground state calculation employing the AdaptVQE algorithm."""
+#figure out which ansatz is used by qiskit nature
+from typing import Optional, List, Tuple, Union, Callable
+
+import copy
+import re
+import logging
+
+import numpy as np
+from qiskit import QiskitError
+
+from qiskit.algorithms import VQE
+from qiskit.algorithms.minimum_eigen_solvers.vqe import VQEResult
+from qiskit.circuit import QuantumCircuit
+from qiskit.opflow import OperatorBase, PauliSumOp, ExpectationBase, CircuitSampler
+from qiskit.opflow.gradients import GradientBase, Gradient
+from qiskit.algorithms.minimum_eigen_solvers import MinimumEigensolver
+from qiskit.circuit.library import EvolvedOperatorAnsatz, RealAmplitudes,PauliEvolutionGate
+from qiskit.circuit.library import NLocal
+from qiskit.utils.validation import validate_min
+from qiskit_nature import ListOrDictType
+from qiskit_nature.algorithms.ground_state_solvers.minimum_eigensolver_factories.minimum_eigensolver_factory import MinimumEigensolverFactory
+from qiskit_nature.exceptions import QiskitNatureError
+from qiskit_nature.circuit.library import UCC
+from qiskit_nature.operators.second_quantization import SecondQuantizedOp
+from qiskit_nature.converters.second_quantization import QubitConverter
+from qiskit_nature.converters.second_quantization.utils import ListOrDict
+from qiskit_nature.problems.second_quantization import BaseProblem
+from qiskit_nature.results import ElectronicStructureResult
+from qiskit_nature.deprecation import deprecate_arguments
+from qiskit.algorithms.optimizers import Optimizer
+from qiskit.utils import QuantumInstance
+from qiskit.providers import BaseBackend
+from qiskit.providers import Backend
+
+#from .minimum_eigensolver_factories import MinimumEigensolverFactory
+#from .ground_state_eigensolver import GroundStateEigensolver
+
+logger = logging.getLogger(__name__)
+
+
+class AdaptVQE2(VQE):
+    """A ground state calculation employing the AdaptVQE algorithm.
+
+    The performance of AdaptVQE can significantly depend on the choice of gradient method, QFI
+    solver (if applicable) and the epsilon value.
+
+    To reproduce the default behavior of AdaptVQE prior to Qiskit Nature 0.4 you should supply
+    `delta=1` explicitly. This will use a finite difference scheme for the gradient evaluation
+    whereas after version 0.4 a parameter shift gradient will be used.
+    [https://qiskit.org/documentation/tutorials/operators/02_gradients_framework.html]
+    """
+
+    @deprecate_arguments(
+        "0.4.0",
+        {"delta": "gradient"},
+        additional_msg=(
+            "Instead of `delta=1.0` you have to construct a gradient, like so "
+            "`gradient=Gradient(grad_method='fin_diff', epsilon=1.0)`."
+        ),
+    )
```

Comment:
All this deprecation aspect can be completely removed. It will be a new algorithm in Terra (and as such at some point a reno release note will be needed in this PR as well)